### DISCUSSION TITLE
Use Cake.Frosten without command line parsing...

### DISCUSSION THREAD

**Discussion by tapika** [cake-build/cake#4067] [Ideas]
Hi !

I think it would be nice to use `Cake.Frosten` without need to parse command line arguments.

Maybe syntax could be something like 
```
IFrostingContext context = new CakeHost().Context;
context.DotNetRestore(...);
```

**Comment by gep13** [comment]
@tapika I am not sure I follow what you are asking for here.  In the example that you provide, I don't see any command line argument parsing happening. Can you elaborate on what exactly you would like to see, perhaps comparing what you currently have to do.

Here, I am assuming you are referring to Cake.Frosting, as there isn't anything called Cake.Frosten, unless you are referring to a 3rd party addin.

**Comment by devlead** [comment]
> Hi !
> 
> I think it would be nice to use `Cake.Frosten` without need to parse command line arguments.
> 
> Maybe syntax could be something like
> 
> ```
> IFrostingContext context = new CakeHost().Context;
> context.DotNetRestore(...);
> ```

`Cake.Frosting` is an opinionated alternative to Cake scripting experience. By you're example code it looks more like you're just looking for a way to get an `ICakeContext` so you can use aliases on it. I'm wondering if it's something like [Cake.Bridge.DependencyInjection](https://www.nuget.org/packages/Cake.Bridge.DependencyInjection) you're after.

Using it would look something like this
```csharp
using Cake.Bridge.DependencyInjection;
using Cake.Common.Tools.DotNet;
using Cake.Core;
using Microsoft.Extensions.DependencyInjection;

var serviceCollection = new ServiceCollection()
    .AddCakeCore();

var context = serviceCollection
                            .BuildServiceProvider()
                            .GetRequiredService<ICakeContext>();

context.DotNetRestore();
```

Example gist of above: https://gist.github.com/devlead/5cff67bf9aa87e839b97d46dca5a5f1f#file-program-cs

**Reply by tapika** [reply]
Yes, that exactly what I wanted to get. I've sketched somehow ugly hack like below, but will switch to your approach indeed, as it looks better.

```
using Cake.Core;
using Cake.Frosting;
using Microsoft.Extensions.DependencyInjection;

public class BuildContext : FrostingContext
{
    static CakeHost _host;

    public static CakeHost Host
    {
        get { 
            if(_host == null )
                _host = new CakeHost().UseContext<BuildContext>();
            return _host;
        }
    }
        
    static ICakeContext _context = null;

    public static ICakeContext Context
    {
        get {
            if (_context == null)
            {
                var h = Host;
                h.ConfigureServices((config) =>
                {
                    config.AddSingleton<ICakeReportPrinter,NullRenderer>();
                });
                h.Run(new[] { "-t", "cake_router_task", "-v", "quiet" });
            }

            return _context;
        }

        set {
            _context = value;
        }
    }

    public BuildContext() : base(Context)
    { 
    }

    public BuildContext(ICakeContext context) : base(context)
    {
    }
}

[TaskName(nameof(cake_router_task))]
public class cake_router_task : FrostingTask
{
    public override void Run(ICakeContext context)
    {
        BuildContext.Context = context;
    }
}

class NullRenderer : ICakeReportPrinter
{
    public void Write(CakeReport report)
    {
    }
}
```

and suspect also slower.
